File: 408-valid-word-abbreviation/valid-word-abbreviation.java
class Solution {
    public boolean validWordAbbreviation(String word, String abbr) {
        int wordInd = 0, abbrInd = 0, currNum = 0;
        while(abbrInd < abbr.length()) {
            char c = abbr.charAt(abbrInd);
            if(Character.isDigit(c)) {
                if(abbr.charAt(abbrInd++) == '0' && currNum == 0) return false;
                currNum*= 10;
                currNum+= c - '0';
            }else {
                wordInd+= currNum;
                currNum = 0;
                if(wordInd >= word.length() || word.charAt(wordInd++) != abbr.charAt(abbrInd++)) return false;
            }
        }
        wordInd+= currNum;
        return wordInd == word.length();
    }
}

File: 408-valid-word-abbreviation/valid-word-abbreviation.cpp
class Solution {
public:
    bool validWordAbbreviation(string word, string abbr) {
        int wordInd = 0, abbrInd = 0, currNum = 0;
        while(abbrInd < abbr.size()){
            if(isdigit(abbr.at(abbrInd))){
                if(currNum == 0 && abbr.at(abbrInd) == '0') return false;
                currNum*= 10;
                currNum+= (int)(abbr.at(abbrInd) - '0');
            }else{
                while(currNum-- > 0){
                    wordInd++;
                    
                }
                currNum = 0;
                if(wordInd >= word.size() || word.at(wordInd++) != abbr.at(abbrInd)) return false;

            }
            abbrInd++;
        }
        while(currNum-- > 0) wordInd++;
        return wordInd == word.size();
    }
};